include/openmc/particle_data.h

```diff
@@ -0,0 +1,453 @@
+  int index_temp_sab;   //!< Temperature index for sab_tables
+  double sab_frac;      //!< Fraction of atoms affected by S(a,b)
+  bool use_ptable;      //!< In URR range with probability tables?
+
+  // Energy and temperature last used to evaluate these cross sections.  If
+  // these values have changed, then the cross sections must be re-evaluated.
+  double last_E {0.0};      //!< Last evaluated energy
+  double last_sqrtkT {0.0}; //!< Last temperature in sqrt(Boltzmann constant
+                            //!< * temperature (eV))
+};
+
+//==============================================================================
+//! Cached microscopic photon cross sections for a particular element at the
+//! current energy
+//==============================================================================
+
+struct ElementMicroXS {
+  int index_grid;         //!< index on element energy grid
+  double last_E {0.0};    //!< last evaluated energy in [eV]
+  double interp_factor;   //!< interpolation factor on energy grid
+  double total;           //!< microscopic total photon xs
+  double coherent;        //!< microscopic coherent xs
+  double incoherent;      //!< microscopic incoherent xs
+  double photoelectric;   //!< microscopic photoelectric xs
+  double pair_production; //!< microscopic pair production xs
+};
+
+//==============================================================================
+// MacroXS contains cached macroscopic cross sections for the material a
+// particle is traveling through
+//==============================================================================
+
+struct MacroXS {
+  double total;       //!< macroscopic total xs
+  double absorption;  //!< macroscopic absorption xs
+  double fission;     //!< macroscopic fission xs
+  double nu_fission;  //!< macroscopic production xs
+  double photon_prod; //!< macroscopic photon production xs
+
+  // Photon cross sections
+  double coherent;        //!< macroscopic coherent xs
+  double incoherent;      //!< macroscopic incoherent xs
+  double photoelectric;   //!< macroscopic photoelectric xs
+  double pair_production; //!< macroscopic pair production xs
+};
+
+//==============================================================================
+// Information about nearest boundary crossing
+//==============================================================================
+
+struct BoundaryInfo {
+  double distance {INFINITY}; //!< distance to nearest boundary
+  int surface_index {0}; //!< if boundary is surface, index in surfaces vector
+  int coord_level;       //!< coordinate level after crossing boundary
+  array<int, 3>
+    lattice_translation {}; //!< which way lattice indices will change
+};
+
+//============================================================================
+//! Defines how particle data is laid out in memory
+//============================================================================
+
+class ParticleData {
+
+public:
+  ParticleData();
+
+private:
```

Comment:
Maybe we should make this `protected:` so that `Particle` still has access to all the members. The thing I don't like about calling, e.g., `alive()`, `material()`, etc. in the methods of `Particle` is that it becomes less clear whether those are accessor methods of the `Particle` class or just external functions. Using the data members directly (`alive_`, `material_`, etc.) immediately signals to the reader that those are data members of the current class. The alternative is to put `this->` in front of everything which gets busy (not in favor of that).

Reply:
Totally agree with you that this is less clear. Normally, I would not design a class to be using accessor methods to variables it owns either.

But, this ambiguity is the very reason for putting these as methods. In my structure-of-array code, you call `material()` as in the code in this PR, but it's instead going to access an array rather then return a member variable. So, I template on the base class to decide who provides a particle's energy, whether `E()` is defined to simply return a member variable or return a reference to a position in some array. In my view of it, the `Particle` class should do abstracted physical operations on pieces of data like the energy, position, etc. No need to know how the data is laid out in memory, just that it will do certain things with that data when the particle collides.

If we just accessed the member variables here, it would break compatibility with the SOA code, making us have separate `event_collide` and so on between the CPU and GPU code.

This is the best solution I could come up with. If you have other ideas, my ears are totally open. This is the most simple route that avoids having separate copies of the same code.

Reply:
That all makes complete sense @gridley. Thanks for taking the time to elaborate on the design decisions!